mmdet/models/roi_heads/standard_roi_head.py

```diff
@@ -0,0 +1,290 @@
+            sampling_results = []
+            for i in range(num_imgs):
+                assign_result = self.bbox_assigner.assign(
+                    proposal_list[i], gt_bboxes[i], gt_bboxes_ignore[i],
+                    gt_labels[i])
+                sampling_result = self.bbox_sampler.sample(
+                    assign_result,
+                    proposal_list[i],
+                    gt_bboxes[i],
+                    gt_labels[i],
+                    feats=[lvl_feat[i][None] for lvl_feat in x])
+                sampling_results.append(sampling_result)
+
+        losses = dict()
+        # bbox head forward and loss
+        if self.with_bbox:
+            loss_bbox, bbox_feats = self._bbox_forward_train(
+                x, sampling_results, gt_bboxes, gt_labels, img_metas)
+            losses.update(loss_bbox)
+
+        # mask head forward and loss
+        if self.with_mask:
+            loss_mask = self._mask_forward_train(x, sampling_results,
+                                                 bbox_feats, gt_masks,
+                                                 img_metas)
+            if loss_mask is not None:
+                losses.update(loss_mask)
+
+        return losses
+
+    def _bbox_forward(self, x, rois):
+        # TODO: a more flexible way to decide which feature maps to use
+        bbox_feats = self.bbox_roi_extractor(
+            x[:self.bbox_roi_extractor.num_inputs], rois)
+        if self.with_shared_head:
+            bbox_feats = self.shared_head(bbox_feats)
+        cls_score, bbox_pred = self.bbox_head(bbox_feats)
+        return cls_score, bbox_pred, bbox_feats
+
+    def _bbox_forward_train(self, x, sampling_results, gt_bboxes, gt_labels,
+                            img_metas):
+        rois = bbox2roi([res.bboxes for res in sampling_results])
+        cls_score, bbox_pred, bbox_feats = self._bbox_forward(x, rois)
+
+        bbox_targets = self.bbox_head.get_target(sampling_results, gt_bboxes,
+                                                 gt_labels, self.train_cfg)
+        loss_bbox = self.bbox_head.loss(cls_score, bbox_pred, *bbox_targets)
+        return loss_bbox, bbox_feats
+
+    def _mask_forward_train(self, x, sampling_results, bbox_feats, gt_masks,
+                            img_metas):
+        mask_feats = self.extract_mask_feats(x, sampling_results, bbox_feats)
+
+        if mask_feats.shape[0] > 0:
+            mask_pred = self.mask_head(mask_feats)
+            mask_targets = self.mask_head.get_target(sampling_results,
+                                                     gt_masks, self.train_cfg)
+            pos_labels = torch.cat(
+                [res.pos_gt_labels for res in sampling_results])
+            loss_mask = self.mask_head.loss(mask_pred, mask_targets,
+                                            pos_labels)
+            return loss_mask
+        else:
+            return None
+
+    def extract_mask_feats(self, x, sampling_results, bbox_feats):
```

Comment:
Use unified apis. There are `_bbox_forward_train` and `_bbox_forward`, but `_mask_forward_train` and `extract_mask_feats`.